### PR TITLE
[mssql] add typing for connect function

### PR DESCRIPTION
**Pull request by MasterOdin**
Signed-off-by: Matthew Peveler <matt.peveler@gmail.com>

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
The export of connection function happens in https://github.com/tediousjs/node-mssql/blob/master/lib/global-connection.js#L8-L63, which is picked up and re-exported in https://github.com/tediousjs/node-mssql/blob/master/lib/base/index.js#L37, and then surfaced in https://github.com/tediousjs/node-mssql/blob/master/lib/tedious/index.js#L13, and finally to general consumption in https://github.com/tediousjs/node-mssql/blob/master/index.js#L1. This looks to be added in the 6.0 release.

The first example on the README uses this `connect` function: https://github.com/tediousjs/node-mssql#quick-example

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## mssql ([<kbd>unpkg</kbd>](https://unpkg.com/browse/mssql@latest/))
was missing the following properties:
1. getTypeByValue
1. close
1. on
1. removeListener
1. off

<details>
<summary>as well as these 2 other properties...</summary>
<p>query, batch</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 954e16b65087243d735024fb3d44a5026b0f6fec
</p>

File: types/mssql/index.d.ts
Comment (achrinza) [inline types/mssql/index.d.ts:398]: 1. While we're at it, can we add typings where `connect()` accepts 0 arguments if there's an existing global connection?

    **References:**
    - https://github.com/tediousjs/node-mssql/blob/89492de2227f0876100ecc7af2f5d352c348d4ab/lib/global-connection.js#L17
    - https://github.com/tediousjs/node-mssql/blame/89492de2227f0876100ecc7af2f5d352c348d4ab/README.md#L358-L361

2. Perhaps we can also consider merging and simplifying the type definitions? IMO this would improve readability,  and is what's followed by the existing types:

    <details>
    <summary>Existing examples</summary>

    https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f3130558c77ec1d970840c9451d49c20b20f81fa/types/mssql/index.d.ts#L259
    https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f3130558c77ec1d970840c9451d49c20b20f81fa/types/mssql/index.d.ts#L283
    https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f3130558c77ec1d970840c9451d49c20b20f81fa/types/mssql/index.d.ts#L322-L323
    https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f3130558c77ec1d970840c9451d49c20b20f81fa/types/mssql/index.d.ts#L362-L363
    https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f3130558c77ec1d970840c9451d49c20b20f81fa/types/mssql/index.d.ts#L381-L382

    </details>

```suggestion
export declare function connect(config?: config | string, callback: (err?: any, pool?: ConnectionPool) => void): void;
export declare function connect(config?: config | string): Promise<ConnectionPool>;
```

**Review by achrinza — COMMENTED** [review]
Thanks for adding the type definitions; Just a few non-blocking comments 👇 

I'd also like to get the opinion of another definition maintainer before merging.

File: types/mssql/index.d.ts
Comment (MasterOdin) [reply types/mssql/index.d.ts]: I applied your first point.

On the second one, I copied the constructor signature for `Connection` class which is what this function hooks into:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bab884457e31721eb800af45ca9a013fb0054d63/types/mssql/index.d.ts#L222-L223

It would then probably make sense to consolidate those as well then?

File: types/mssql/index.d.ts
Comment (achrinza) [inline types/mssql/index.d.ts:396]: Sorry for missing the updates.

Can we keep the `config` type consistent?

```suggestion
export declare function connect(config?: string | config, callback: (err?: any, pool?: ConnectionPool) => void): void;
export declare function connect(config?: string | config): Promise<ConnectionPool>;
```

**Review by achrinza — CHANGES_REQUESTED** [review]

File: types/mssql/mssql-tests.ts
Comment (elhaard) [inline types/mssql/mssql-tests.ts:57]: While this makes sense as a test, the semantics of it makes my head explode :-)

**Review by elhaard — COMMENTED** [review]
Thank you for making this PR.

The [Quick Exaple](https://github.com/tediousjs/node-mssql#quick-example) also mentions the global `query()`, so it would be great if you could add that as well, along with `batch()` and  `close()`

DangerJS also notices that `getTypeByValue()`, `on()`, `off()` and `removeListener()` are missing. They are probably slightly less used and could easily wait to another PR. (But extra doubleplusgood if you could add them, too :-) ).

**Pull request closed without merging**
